File: dirShort.py
#!/usr/bin/python3
# ##############################################################################
# dirShort.py
# A script that allows the working directory string length to be dynamically
# shortened 
#
# Author: Craig B. Rafter (https://github.com/cbrafter)
# Copyright: 2017, Craig Rafter, MIT License
# Date: 28/02/17
# Useage: Add following to ~/.bashrc and place this script in the home directory
# ensure you give the script any relevant permissions (chmod)
#   PS1='\[\033[01;32m\]\u@\h\[\033[00m\]:\[\033[01;34m\]$(python3 $HOME/dirShort.py "$(dirs)")\[\033[00m\]\$ '
# ##############################################################################

import sys
from os import popen

try:
    # system provides working directory
    path = sys.argv[1]
except:
    print("ERROR: Usage: $0 <path>")
    sys.exit(1)

# Separate folders
dirs = path.split("/");

# Width of terminal window in characters
termWidth = int(popen('stty size', 'r').read().split()[1])
cutoff = 0.4*termWidth # 40% window size rounded up to nearest int

# If within length, pass 
if len(path) <= cutoff:
    pass
# If more than 1 folder deep
elif len(dirs) > 1:
    # Reduce num letters in parent folder until 1 letter per parent to 
    # try fit the dir in the cutoff area
    for parentLen in [5, 3, 1]:
        path = '/'.join(x[:parentLen] for x in dirs[:-1]) + '/' + dirs[-1]
        if len(path) <= cutoff:
            break
# If there's a problem just return the directory
else:
    path = dirs[0]

print(path)

File: dirShort.c
/* dirShort.c

A program that allows the working directory string length to be dynamically
shortened

Author: Craig B. Rafter (https://github.com/cbrafter)
Copyright: 2017, Craig Rafter, MIT License
Date: 28/02/17
Useage: Compile this file using:

	gcc -O3 -Wall -std=c99 dirShort.c -o dirShort 

place the executable in the home directory
ensure you give the executeable any relevant permissions (chmod)

Then add following to ~/.bashrc:

	PS1='\[\033[01;32m\]\u@\h\[\033[00m\]:\[\033[01;34m\]$($HOME/dirShort "$(dirs)")\[\033[00m\]\$ '
 */
#include<stdio.h>
#include<stdlib.h>
#include<string.h>
#include<sys/ioctl.h>

#define max(a,b) \
   ({ __typeof__ (a) _a = (a); \
       __typeof__ (b) _b = (b); \
     _a > _b ? _a : _b; })


int main(int argc, char **argv){
	/*Check only two inputs*/
	if(argc != 2){
		printf("Error incorrect arguments passed");
		return -1;
	}

	/*Variables*/
	char *path;
	char *newPath;
	size_t pathLen = strlen(argv[1]);
	/*Copy argv[1] i.e. path for use - strtok edits so copy needed */
	path = (char*)malloc(pathLen * sizeof(char));
	newPath = (char*)malloc(pathLen * sizeof(char));
	strcpy(path, argv[1]);

	/*store directory tokens and number of tokens*/
	char dirs[32][256];
	const char delim[] = "/";
	unsigned short ntoken = 0;
	
	struct winsize window; /*Window handle*/
    ioctl(0, TIOCGWINSZ, &window); /*Get window size object*/
   	unsigned short cutoff = max(15, 0.4*window.ws_col - 15); /*point at which file path
   	shouldn't go past*/

	/*Tokenise string into dirs*/
    char *token = strtok(path, delim);
    while(token){
    	strcpy(dirs[ntoken], token);
    	ntoken++;
        token = strtok(NULL, delim);
    }

	if(pathLen < cutoff){
		printf("%s\n", argv[1]);
	}
	else if(ntoken > 1){
		unsigned short i, j, terminator;
		
		/*Leading path characters*/
		if(argv[1][0] == '/'){
			sprintf(newPath, "/");
			terminator = 1;
		}
		else{
			terminator = 0;
		}

		for(i=9; i>0; i-=2){
			/*reterminate string on each pass to prevent the new strings being 
			added on to the old one*/
			newPath[terminator] = '\0';
			for(j=0; j<ntoken-1; j++){
				strncat(newPath, dirs[j], i); /*truncate parent folders to i chars*/
				strcat(newPath, "/"); /*add / between folders except last*/
			}
			/*Add final folder in full*/
			strcat(newPath, dirs[ntoken-1]);
			/*if the string is sufficiently shortened then stop*/
			if(strlen(newPath) < cutoff)
				break;
		}
		printf("%s\n", newPath);
	}
	else{
		printf("%s\n", argv[1]);
	}

	return 0;
}